contracts/Staking.sol

```diff
@@ -0,0 +1,576 @@
+    require(msg.value >= oracleRelayerFee, "not enough relay fee");
+    if (amount < minDelegation) {
+      require(amount == delegatedOfValidator[msg.sender][validator], "invalid amount");
+    }
+    require(block.timestamp >= pendingUndelegateTime[msg.sender][validator], "pending undelegation exist");
+    delegatedOfValidator[msg.sender][validator] = delegatedOfValidator[msg.sender][validator].sub(amount, "not enough funds");
+
+    uint256 convertedAmount = amount.div(TEN_DECIMALS); // native bnb decimals is 8 on BBC, while the native bnb decimals on BSC is 18
+    uint256 _oracleRelayerFee = msg.value;
+
+    bytes[] memory elements = new bytes[](3);
+    elements[0] = msg.sender.encodeAddress();
+    elements[1] = validator.encodeAddress();
+    elements[2] = convertedAmount.encodeUint();
+    bytes memory msgBytes = elements.encodeList();
+    ICrossChain(CROSS_CHAIN_CONTRACT_ADDR).sendSynPackage(CROSS_STAKE_CHANNELID, _RLPEncode(EVENT_UNDELEGATE, msgBytes), _oracleRelayerFee.div(TEN_DECIMALS));
+    payable(TOKEN_HUB_ADDR).transfer(_oracleRelayerFee);
+
+    delegated[msg.sender] = delegated[msg.sender].sub(amount);
+    pendingUndelegateTime[msg.sender][validator] = block.timestamp.add(8*24*3600);
+
+    emit undelegateSubmitted(msg.sender, validator, amount, _oracleRelayerFee);
+  }
+
+  function redelegate(address validatorSrc, address validatorDst, uint256 amount) override external payable tenDecimalPrecision(amount) initParams {
+    require(validatorSrc != validatorDst, "invalid redelegation");
+    require(msg.value >= oracleRelayerFee, "not enough relay fee");
+    require(amount >= minDelegation, "invalid amount");
+    require(block.timestamp >= pendingRedelegateTime[msg.sender][validatorSrc][validatorDst] &&
+      block.timestamp >= pendingRedelegateTime[msg.sender][validatorDst][validatorSrc], "pending redelegation exist");
+    delegatedOfValidator[msg.sender][validatorSrc] = delegatedOfValidator[msg.sender][validatorSrc].sub(amount, "not enough funds");
+    delegatedOfValidator[msg.sender][validatorDst] = delegatedOfValidator[msg.sender][validatorDst].add(amount);
+
+    uint256 convertedAmount = amount.div(TEN_DECIMALS);// native bnb decimals is 8 on BBC, while the native bnb decimals on BSC is 18
+    uint256 _oracleRelayerFee = msg.value;
+
+    bytes[] memory elements = new bytes[](4);
+    elements[0] = msg.sender.encodeAddress();
+    elements[1] = validatorSrc.encodeAddress();
+    elements[2] = validatorDst.encodeAddress();
+    elements[3] = convertedAmount.encodeUint();
+    bytes memory msgBytes = elements.encodeList();
+    ICrossChain(CROSS_CHAIN_CONTRACT_ADDR).sendSynPackage(CROSS_STAKE_CHANNELID, _RLPEncode(EVENT_REDELEGATE, msgBytes), _oracleRelayerFee.div(TEN_DECIMALS));
+    payable(TOKEN_HUB_ADDR).transfer(_oracleRelayerFee);
+
+    pendingRedelegateTime[msg.sender][validatorSrc][validatorDst] = block.timestamp.add(8*24*3600);
```

Comment:
Can we use `691200` to save the multiply operation here to save gas?

Reply:
Done